compile with {fmt} v8

tchaikov opened the pull request:
## Cover letter

now that {fmt} v8 is released. and would be great if we can compile against it. in this changeset

- since fmt v8 differentiate compile-time fmt string from the runt-time fmt string. the former is a static one and can be validated at compile-time, the later can only be validated at runt-time. so in this change, the format string are passed in different ways accordingly for better performance and compile-time validation.
- to simplify the handling of difference API of v8 and pre-v8, fmt is also packaged as yet another dependency
- fmt v8 does not format scoped enums. so add a bunch of `fmt::format()` specialization for better debugging experience.

<!-- Use the GitHub keyword `Fixes` to link to bug(s) this PR will fix. -->
Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

## Release notes

* none

tchaikov commented on src/v/ssx/sformat.h:
```diff
@@ -24,6 +24,18 @@ struct fmt::formatter<seastar::sstring> {
     }
 };
 
+#if FMT_VERSION >= 80000
```
not sure if `ssx/sformat.h` is the best place for this wrapper. i assume "ssx" is short for seastar extension, but `fmt_append()` is but a wrapper of `fmt::format_to()`, which is not directly connected with Seastar..

dotnwat replied:
>  i assume "ssx" is short for seastar extension

yes that's right.

> not sure if ssx/sformat.h is the best place for this wrapper.

I'll let Ben chime in here, but I think you are right. My guess is that these are only used by `ssx::sformat` (specialization for `seastar::sstring`) and if we had wider usage in the tree we'd move this to a more generic location.

tchaikov replied:
> > not sure if ssx/sformat.h is the best place for this wrapper.
> 
> I'll let Ben chime in here, but I think you are right. My guess is that these are only used by `ssx::sformat` (specialization for `seastar::sstring`) and if we had wider usage in the tree we'd move this to a more generic location.

i am afraid this is not quite right. i am trying to address the FTBFS caused by `fmt::format_to(format_str, ...)` where `format_str` is not constexpr. and we do have a handful of callers which simply pass the `format_str`to the underlying `fmt::format_to()`. they are using `fmt::format_to()` directly. even the `format_str` is a constexpr when it gets passed to these callers. but the `constexpr`ness is erased when it is passed down to `fmt::format_to()`, hence the error.

tchaikov replied:
i will take a look at the changelog/git history of {fmt} to find out the minimum version where `fmt::format_to(std::back_inserter(buf), format_str, std::forward<Args>(args)...);` works in hope to drop the deprecated version. but i am not quite optimistic in this regard.. as Seastar claims to support fmt 5.0.0. see https://github.com/scylladb/seastar/blob/acf7e3523ba7e20d1bfb09c92a705dcc3e7f0161/cmake/SeastarDependencies.cmake#L87 .

dotnwat replied:
Thanks Kefu! @BenPope could take a look? This is probably something we want to address now or right after we cut our next release and bump clang and seastar and other dependency versions?

dotnwat commented:
:wave: Kefu!

dotnwat commented:
@tchaikov there is a `.clang-format` in the root. looks like the files you touched may need run through clang format (CI is complaining here: https://github.com/redpanda-data/redpanda/runs/5990346426?check_suite_focus=true)

BenPope commented:
I took a little look at updating libfmt a while ago, but didn't have the bandwidth to figure out what the internal changes are. 

I think we should wait until we branch v21.1.x to merge (that doesn't need to block changes on this PR).

libfmt-devel >=8 isn't in Ubuntu until Jammy (April 21st) and Fedora 35, so we may also need to package it in the oss build.

tchaikov commented:
> @tchaikov there is a `.clang-format` in the root of the tree. looks like the files you touched may need run through clang format (CI is complaining here: https://github.com/redpanda-data/redpanda/runs/5990346426?check_suite_focus=true)

thanks for the pointer. sure, will do.

senior7515 commented:
good to see you here @tchaikov 👋🏽 🚀

tchaikov commented:
> libfmt-devel >=8 isn't in Ubuntu until Jammy (April 21st) and Fedora 35, so we may also need to package it in [the oss build](https://github.com/tchaikov/redpanda/blob/188b601fadf1d2b8cb4d12836b0efd2a27a302a1/cmake/oss.cmake.in).

okay, will add a commit in this PR to package the latest stable version of fmt in `oss.cmake.in` . much simpler this way.

tchaikov commented:
@BenPope hi Ben, thanks for your review. i just sent all changes related to fmt v8. now the tree compiles just fine with them and some other misc changes (also proposed as PRs). but i will have two questions. probably you could shed some lights on them?

- shall i drop the backward compatible implementation for `fmt_append()`
- shall i keep `fmt_append()` in `src/v/ssx/sformat.h`? if not, would you please suggest where i shall put it?

jcsp commented:
Ping @BenPope I think this is waiting for your input?

BenPope commented:
Thanks @jcsp, I'm aware.

BenPope commented:
@tchaikov - thanks for this. I hacked on this a bit - my branch is it: https://github.com/BenPope/redpanda/commits/update-fmt-v8

It could do with some tidy up of the history. I spent quite a while on a big performance regression.

tchaikov commented:
@BenPope much appreciated! i will fold your changes into my branch with your credits.

tchaikov commented:
hi Ben, in the latest revision, i 

- incorporated all your fmt-v8 related change in https://github.com/BenPope/redpanda/commits/update-fmt-v8 
- dropped my change introducing `fmt_append()` in favor of your sformat change.
- folded your ["[differentiate compile-time fmt str from runtime str](https://github.com/redpanda-data/redpanda/pull/4260/commits/8b7dde5fbfc9e8649b4bf5959c82fca3769e4c58)" ](https://github.com/BenPope/redpanda/commit/5b0d1d7080288c8c6afa826057719a60352c50d7) into the "differentiate compile-time fmt str..." change
- added 8b7dde5fbfc9e8649b4bf5959c82fca3769e4c58 to address a new FTBFS

i will squash the related change into a single one after the review.

BenPope commented:
I'm seeing a compilation failure:
```
In file included from ../../../src/v/pandaproxy/rest/handlers.cc:21:
  | In file included from ../../../src/v/pandaproxy/json/exceptions.h:15:
  | ../../../src/v/pandaproxy/error.h:73:8: error: explicit specialization of 'std::is_error_code_enum<kafka::error_code>' after instantiation
  | struct is_error_code_enum<kafka::error_code> : true_type {};
  | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  | /vectorized/llvm/bin/../include/c++/v1/system_error:328:39: note: implicit instantiation first required here
  | typename enable_if<is_error_code_enum<_Ep>::value>::type* = nullptr
```

tchaikov commented:
ahh, this failure is one of the day0 failures i wanted to address in #4307 which should be superseded by #4722.

BenPope commented:
@tchaikov can you rebase this on dev, #4722 has been merged.

tchaikov commented:
@BenPope sure. squashed, rebased and repushed.

dotnwat commented:
retriggered internal ci

tchaikov commented:
change since previous revision

- clang-fmt'ed

BenPope reviewed (commented):
LGTM.

Out internal build and tests have passed.

This will have to be squashed, as it's not worth trying to make all intermediate steps compatible with both v7 and v8, and it will make git bisect easier.

tchaikov commented on src/v/rpc/demo/demo_utils.h:
```diff
@@ -14,7 +14,9 @@
 #include "seastarx.h"
 #include "utils/hdr_hist.h"
-
+// clang-format off
+#include "ssx/sformat.h"
+// clang-format on
```
IIRC, i was trying to avoid an instantiated-before-definition error. will recompile the tree to reproduce it and paste the error message.

tchaikov replied:
ahh! i found it. actually i pasted in the commit message:
```
    to get the fmt::formatter<seastar::sstring> instantiated before
    it is implicit instantiated by Seastar library.

    this change helps to address the FTBFS of

    /usr/bin/clang++-15 -DFMT_LOCALE -DFMT_SHARED -DSEASTAR_API_LEVEL=6 -DSEASTAR_DEBUG -DSEASTAR_DEBUG_SHARED_PTR -DSEASTAR_DEFAULT_ALLOCATOR -DSEASTAR_SCHEDULING_GROUPS_COUNT=16 -DSEASTAR_SHUFFLE_TASK_QUEUE -DSEASTAR_TYPE_ERASE_MORE -DXXH_PRIVATE_API -DZSTD_STATIC_LINKING_ONLY -I/home/kefu/dev/redpanda/src/v -I/home/kefu/dev/redpanda/build/src/v -isystem /home/kefu/dev/redpanda/build/deps_install/include -isystem /home/kefu/dev/redpanda/build/deps_install/include/hdr -fPIC -fsanitize=undefined -fsanitize=address -fcolor-diagnostics -g -fPIC -Wall -Wextra -Werror -Wno-missing-field-initializers -std=c++20 -U_FORTIFY_SOURCE -DSEASTAR_SSTRING -Werror=unused-result "-Wno-error=#warnings" -fstack-clash-protection -fsanitize=address -fsanitize=undefined -fno-sanitize=vptr -std=c++20 -MD -MT src/v/rpc/CMakeFiles/v_rpc.dir/reconnect_transport.cc.o -MF src/v/rpc/CMakeFiles/v_rpc.dir/reconnect_transport.cc.o.d -o src/v/rpc/CMakeFiles/v_rpc.dir/reconnect_transport.cc.o -c /home/kefu/dev/redpanda/src/v/rpc/reconnect_transport.cc
    In file included from /home/kefu/dev/redpanda/src/v/rpc/reconnect_transport.cc:13:
    In file included from /home/kefu/dev/redpanda/src/v/raft/logger.h:14:
    In file included from /home/kefu/dev/redpanda/src/v/model/fundamental.h:15:
    /home/kefu/dev/redpanda/src/v/ssx/sformat.h:19:13: error: explicit specialization of 'fmt::formatter<seastar::basic_sstring<char, unsigned int, 15, true>>' after instantiation
    struct fmt::formatter<seastar::sstring> {
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~
    /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/type_traits:971:30: note: implicit instantiation first required here
        : public __bool_constant<__is_constructible(_Tp, _Args...)>
                                 ^
```

tchaikov commented on src/v/model/metadata.h:
```diff
@@ -351,6 +351,25 @@ struct broker_v0 {
+    constexpr auto parse(format_parse_context& ctx) { return ctx.end(); }
+    template<typename FormatContext>
+    auto format(const isolation_level& s, FormatContext& ctx) const {
+        const char* str = "unknown";
```
ack.

tchaikov commented:
change since the previous version

- squashed into a single commit. the commit messages are itemized and merged.
- addressed the `string_view` comments

BenPope reviewed (commented):
https://buildkite.com/redpanda/vtools/builds/2255

BenPope reviewed (approved):
LGTM - thanks @tchaikov !

tchaikov commented:
shall i worry about the test failures? after registering on buildkite, i still have 404 when trying to read any of them.
